tensorboard/plugins/metrics/metrics_plugin.py

```diff
@@ -321,6 +321,11 @@ def _tags_impl(self, ctx, experiment=None):
             experiment_id=experiment,
             plugin_name=image_metadata.PLUGIN_NAME,
         )
+        # Not all data providers support tensors and/or blob sequences.
+        if histogram_mapping is None:
+            histogram_mapping = {}
+        if image_mapping is None:
+            image_mapping = {}
```

Comment:
If unimplemented DataProvider methods may return None, can we add the corresponding `if scalar_mapping is None` check as well, so that metrics_plugin.py doesn't rely on knowing which specific dataclasses have implemented `list_*`?

On the other hand, should `GrpcDataProvider` be responsible for implementing `list_tensors` that returns `{}`?  [1] suggests that instances of classes with `@abc.abstractmethod` must have "all of its abstract methods and properties are overridden".  Would it make sense to add GrpcDataProvider.list_tensors/blobs, or update the `data/provider.py` docstrings to indicate that they may return `None`?

[1] https://docs.python.org/3/library/abc.html#abc.abstractmethod

Reply:
The scalars methods are `abc.abstractmethods` whereas the methods for
other data classes are not. You cannot instantiate a `DataProvider`
without implementing `read_scalars`. It’ll hard-error.

That these methods pass when not implemented is (in my experience)
fairly standard, but I’m happy to document it explicitly.

Reply:
> The scalars methods are abc.abstractmethods whereas the methods for
other data classes are not

Ahh, thanks for pointing this out.  I think my comment earlier was invalid, so feel free to keep or remove the data/provider.py change as you see fit!